### PR TITLE
[SEMI-MODULAR] Adds a small icon/examine blurb for people with DNR to med/sec huds

### PR DESCRIPTION
**Pull request by nikothedude**
## About The Pull Request

Title. Its just a tiny lil red/white exclamation mark and a blurb in warning span in the guy isnt already dead.

## How This Contributes To The Skyrat Roleplay Experience

A common complaint about people in this game is how people always say "ahahah nobody can die ahhah just kill them and revive them ahaha"

Theyre not WRONG, but it does cause some issues in scenarios where its NOT true, namely in the DNR quirk. Like with my temp ftext icon, having this info available on-a-glance and clarifying when curious people examine will allow a lot more people to recognize this little thing.

Its locked to huds over concerns of cluttering the space around people with icons.

Finally, and I mean finally, we can actually genuinely yell at security who goes "im just gonna shoot the hostage taker" in the MOMENT and PREVENT IT FROM HAPPENING - because we can all see the lil !

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>

image

</details>

## Changelog

:cl:
add: Sec/Medhuds can now see a small ! if a person has the DNR quick, as well as see a blurb in examine about it
/:cl:

### CODE REVIEW

**Comment by nikothedude** [comment]
postponed til i can see through my tears of having to deal with hud code ohhhhhh

**Comment by nikothedude** [comment]
oh that was fucking miserable

**Comment by Namwty** [comment]
Suggestion:
Have more RP friendly text in examine, especially as it's tied to sec/med huds. "...gone for the rest of the **round**" just ain't it, in my opinion.

Issue:

DNR quirk should not give the player special privileges or anything that would resemble one, same with quirks like blood deficiency, fragility, frail and so on. Security shouldn't treat someone who has DNR any different than someone who doesn't, otherwise we will have this issue where DNR people get special treatment, cuz of bold text, and rest people will just get "Hey, this guy doesn't have DNR, let's go ham on them", worst is people who don't have DNR will suffer in hostage situations. Ideally everyone should just treat every death as potential DNR, and people with DNR should not do "suicidal" options, like using lethals on security and then getting blasted.

Now comes the bigger issue, if this quirk get's into ahelps, as a tool to get people punished, it will become a big problem. I could go into deep detail, but I will just have a small example with other quirk:

Person A has 2 quirks - Frail ( wound dmg multiplier ) and Blood deficiency ( low blood level, bordering suffocation dmg ) -> Person A starts shit, which ends Person B ( most likely security ) having to use force -> Person A is literally designed by quirks to be easy to lethal and thus get's killed even in situations where normal crew member would have not died ( most likely bleeding ) -> Person A and some random people in dead chat get mad and salty for Person B over escalating - Person B get's bwoiked for "overescalation" ( as if it's his fault that you could behead Person A with a damn butter knife ).

This shit has happened and most likely still happens ( RARELY, but still ), and I will open a secret, most staff members DO NOT check quirks of people and it's generally not something most folk pay attention to. ( Not a shot at staff as it's time consuming to check and it's not something that even I thought about untill I played shit like ghoul (( has their own "frail/fragility and other" traits )) who literally would get worst wounds from basic crap like, pickaxes, but it's something I rather not become a common problem that antags/security have to face )

So my fear is that DNR will just be tool for ahelping and other crap. Not to mention if Person B hates Person A ( who in this case has DNR ), person B will use knowledge to make bullshit excuse to round remove person. DNR text will just act like a target on their back.

In short, if you die for good cause you have DNR, you shouldn't be able to ahelp them... same with crap like frail/ghouls/blood def, the massive text in examine will just act as enabler for ahelps, as "the person should have known, bad faith from them"


And before people say "We are roleplaying server and people wouldn't abuse and blah blah", there is a reason saying " That's why we can't have nice things" exist and why rules/code restrictions exist.

**Comment by desertmote** [comment]
Great idea, doctors need this info, though security not as much.

File: modular_skyrat/master_files/code/datums/traits/neutral.dm [hunk 1]
```diff
@@ -31,6 +40,41 @@
...
+	if (HAS_TRAIT(src, TRAIT_DNR))
+		set_hud_image_active(DNR_HUD)
+		dnr_holder.icon_state = "hudDNR"
+	else
+		set_hud_image_inactive(DNR_HUD)
+
+/mob/living/carbon/human/examine(mob/user)
+	. = ..()
+
+	if (stat != DEAD && HAS_TRAIT(src, TRAIT_DNR) && (HAS_TRAIT(user, TRAIT_SECURITY_HUD) || HAS_TRAIT(user, TRAIT_MEDICAL_HUD)))
```
Comment (vinylspiders) [inline modular_skyrat/master_files/code/datums/traits/neutral.dm]:
```suggestion
	if(stat != DEAD && HAS_TRAIT(src, TRAIT_DNR) && (HAS_TRAIT(user, TRAIT_SECURITY_HUD) || HAS_TRAIT(user, TRAIT_MEDICAL_HUD)))
```

File: modular_skyrat/master_files/code/datums/traits/neutral.dm [hunk 2]
```diff
@@ -31,6 +40,41 @@
...
+
+/mob/living/proc/update_dnr_hud()
+	var/image/dnr_holder = hud_list?[DNR_HUD]
+	if (isnull(dnr_holder))
+		return
+
+	var/icon/I = icon(icon, icon_state, dir)
+	dnr_holder.pixel_y = I.Height() - world.icon_size
+
+	if (HAS_TRAIT(src, TRAIT_DNR))
```
Comment (vinylspiders) [inline modular_skyrat/master_files/code/datums/traits/neutral.dm]:
```suggestion
	if(HAS_TRAIT(src, TRAIT_DNR))
```

File: modular_skyrat/master_files/code/datums/traits/neutral.dm [hunk 3]
```diff
@@ -31,6 +40,41 @@
...
+/datum/quirk/dnr/remove()
+	var/mob/living/old_holder = quirk_holder
+
+	. = ..()
+
+	old_holder.update_dnr_hud()
+
+/mob/living/proc/update_dnr_hud()
+	var/image/dnr_holder = hud_list?[DNR_HUD]
+	if (isnull(dnr_holder))
```
Comment (vinylspiders) [inline modular_skyrat/master_files/code/datums/traits/neutral.dm]:
```suggestion
	if(isnull(dnr_holder))
```

File: code/modules/mob/living/living.dm [hunk 4]
```diff
@@ -25,6 +25,8 @@
 	med_hud_set_health()
 	med_hud_set_status()
 
+	update_dnr_hud() // SKYRAT EDIT ADDITION - DNR HUD
+
```
Comment (vinylspiders) [inline code/modules/mob/living/living.dm]:
```suggestion
	update_dnr_hud() // SKYRAT EDIT ADDITION - DNR HUD

```

File: code/game/data_huds.dm [hunk 5]
```diff
@@ -47,7 +47,7 @@
 	hud_icons = list(ID_HUD)
 
 /datum/atom_hud/data/human/security/advanced
-	hud_icons = list(ID_HUD, IMPTRACK_HUD, IMPLOYAL_HUD, IMPCHEM_HUD, WANTED_HUD, PERMIT_HUD) //SKYRAT EDIT: ADD PERMIT_HUD
+	hud_icons = list(ID_HUD, IMPTRACK_HUD, IMPLOYAL_HUD, IMPCHEM_HUD, WANTED_HUD, PERMIT_HUD, DNR_HUD) //SKYRAT EDIT: ADD PERMIT_HUD, DNR_HUD
```
Comment (vinylspiders) [inline code/game/data_huds.dm]:
```suggestion
	hud_icons = list(ID_HUD, IMPTRACK_HUD, IMPLOYAL_HUD, IMPCHEM_HUD, WANTED_HUD, PERMIT_HUD, DNR_HUD) //SKYRAT EDIT ADDITION - PERMIT_HUD, DNR_HUD
```

File: code/game/data_huds.dm [hunk 6]
```diff
@@ -18,7 +18,7 @@
 /datum/atom_hud/data
 
 /datum/atom_hud/data/human/medical
-	hud_icons = list(STATUS_HUD, HEALTH_HUD)
+	hud_icons = list(STATUS_HUD, HEALTH_HUD, DNR_HUD) // SKYRAT EDIT ADDITION: DNR_HUD
```
Comment (vinylspiders) [inline code/game/data_huds.dm]:
```suggestion
	hud_icons = list(STATUS_HUD, HEALTH_HUD, DNR_HUD) // SKYRAT EDIT ADDITION - DNR_HUD
```

File: code/datums/hud.dm [hunk 7]
```diff
@@ -19,6 +19,7 @@ GLOBAL_LIST_INIT(huds, list(
 	DATA_HUD_AI_DETECT = new/datum/atom_hud/ai_detector(),
 	DATA_HUD_FAN = new/datum/atom_hud/data/human/fan_hud(),
 	DATA_HUD_PERMIT = new/datum/atom_hud/data/human/permit(), // SKYRAT EDIT
+	DATA_HUD_DNR = new/datum/atom_hud/data/human/dnr(), // SKYRAT EDIT
```
Comment (vinylspiders) [inline code/datums/hud.dm]:
```suggestion
	DATA_HUD_DNR = new/datum/atom_hud/data/human/dnr(), // SKYRAT EDIT ADDITION
```

File: code/datums/hud.dm [hunk 8]
```diff
@@ -19,6 +19,7 @@ GLOBAL_LIST_INIT(huds, list(
 	DATA_HUD_AI_DETECT = new/datum/atom_hud/ai_detector(),
 	DATA_HUD_FAN = new/datum/atom_hud/data/human/fan_hud(),
 	DATA_HUD_PERMIT = new/datum/atom_hud/data/human/permit(), // SKYRAT EDIT
```
Comment (vinylspiders) [inline code/datums/hud.dm]:
```suggestion
	DATA_HUD_PERMIT = new/datum/atom_hud/data/human/permit(), // SKYRAT EDIT ADDITION
```

File: code/__DEFINES/atom_hud.dm [hunk 9]
```diff
@@ -51,6 +51,9 @@
 #define AMMO_HUD "25"
 /// if they have a gun permit
 #define PERMIT_HUD "26"
+
+/// If they have the DNR trait
+#define DNR_HUD "27"
```
Comment (vinylspiders) [inline code/__DEFINES/atom_hud.dm:56]: In line 49 above
`// SKYRAT EDIT ADDITION BEGIN - gun permits`
could you change that to 
`// SKYRAT EDIT ADDITION BEGIN - gun permits and DNR`

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 10]
```diff
@@ -5,7 +5,7 @@
 	icon = 'icons/mob/human/human.dmi'
 	icon_state = "human_basic"
 	appearance_flags = KEEP_TOGETHER|TILE_BOUND|PIXEL_SCALE|LONG_GLIDE
-	hud_possible = list(HEALTH_HUD,STATUS_HUD,ID_HUD,WANTED_HUD,IMPLOYAL_HUD,IMPCHEM_HUD,IMPTRACK_HUD,ANTAG_HUD,GLAND_HUD,SENTIENT_DISEASE_HUD,FAN_HUD,PERMIT_HUD) //SKYRAT EDIT: ADD PERMIT_HUD
+	hud_possible = list(HEALTH_HUD,STATUS_HUD,ID_HUD,WANTED_HUD,IMPLOYAL_HUD,IMPCHEM_HUD,IMPTRACK_HUD,ANTAG_HUD,GLAND_HUD,SENTIENT_DISEASE_HUD,FAN_HUD,PERMIT_HUD,DNR_HUD) //SKYRAT EDIT: ADD PERMIT_HUD, DNR_HUD
```
Comment (vinylspiders) [inline code/modules/mob/living/carbon/human/human_defines.dm]:
```suggestion
	hud_possible = list(HEALTH_HUD,STATUS_HUD,ID_HUD,WANTED_HUD,IMPLOYAL_HUD,IMPCHEM_HUD,IMPTRACK_HUD,ANTAG_HUD,GLAND_HUD,SENTIENT_DISEASE_HUD,FAN_HUD,PERMIT_HUD,DNR_HUD) //SKYRAT EDIT ADDITION - PERMIT_HUD, DNR_HUD
```

File: modular_skyrat/master_files/code/datums/traits/neutral.dm [hunk 11]
```diff
@@ -31,6 +40,41 @@
...
+		return
+
+	var/icon/I = icon(icon, icon_state, dir)
+	dnr_holder.pixel_y = I.Height() - world.icon_size
+
+	if (HAS_TRAIT(src, TRAIT_DNR))
+		set_hud_image_active(DNR_HUD)
+		dnr_holder.icon_state = "hudDNR"
```
Comment (vinylspiders) [inline modular_skyrat/master_files/code/datums/traits/neutral.dm]:
```suggestion
		dnr_holder.icon_state = "hud_dnr"
```

The coding convention for `icon_state` is snake_case (don't forget to change it in the .dmi too)

**Review by vinylspiders — CHANGES_REQUESTED** [review]
These are pretty much all coding conventions/style issues. The code looks solid to me other than that!

File: modular_skyrat/master_files/code/datums/traits/neutral.dm (hunk 3 above)
Comment (nikothedude) [reply modular_skyrat/master_files/code/datums/traits/neutral.dm]: this is convention...? i thought it literally didnt matter

**Review by vinylspiders — APPROVED** [review]

File: code/modules/mob/living/carbon/carbon_defines.dm [hunk 12]
```diff
@@ -2,7 +2,7 @@
 	blood_volume = BLOOD_VOLUME_NORMAL
 	gender = MALE
 	pressure_resistance = 15
-	hud_possible = list(HEALTH_HUD,STATUS_HUD,ANTAG_HUD,GLAND_HUD)
+	hud_possible = list(HEALTH_HUD,STATUS_HUD,ANTAG_HUD,GLAND_HUD,DNR_HUD) // SKYRAT EDIT ADDITION - DNR ICON
```
Comment (GoldenAlpharex) [inline code/modules/mob/living/carbon/carbon_defines.dm]:
```suggestion
	hud_possible = list(HEALTH_HUD,STATUS_HUD,ANTAG_HUD,GLAND_HUD, DNR_HUD) // SKYRAT EDIT ADDITION - DNR ICON
```

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 13]
```diff
@@ -5,7 +5,7 @@
 	icon = 'icons/mob/human/human.dmi'
 	icon_state = "human_basic"
 	appearance_flags = KEEP_TOGETHER|TILE_BOUND|PIXEL_SCALE|LONG_GLIDE
-	hud_possible = list(HEALTH_HUD,STATUS_HUD,ID_HUD,WANTED_HUD,IMPLOYAL_HUD,IMPCHEM_HUD,IMPTRACK_HUD,ANTAG_HUD,GLAND_HUD,SENTIENT_DISEASE_HUD,FAN_HUD,PERMIT_HUD) //SKYRAT EDIT: ADD PERMIT_HUD
+	hud_possible = list(HEALTH_HUD,STATUS_HUD,ID_HUD,WANTED_HUD,IMPLOYAL_HUD,IMPCHEM_HUD,IMPTRACK_HUD,ANTAG_HUD,GLAND_HUD,SENTIENT_DISEASE_HUD,FAN_HUD,PERMIT_HUD,DNR_HUD) //SKYRAT EDIT ADDITION - PERMIT_HUD, DNR_HUD
```
Comment (GoldenAlpharex) [inline code/modules/mob/living/carbon/human/human_defines.dm]: Also for these, you should be adding the original value as per usual.
```suggestion
	hud_possible = list(HEALTH_HUD,STATUS_HUD,ID_HUD,WANTED_HUD,IMPLOYAL_HUD,IMPCHEM_HUD,IMPTRACK_HUD,ANTAG_HUD,GLAND_HUD,SENTIENT_DISEASE_HUD,FAN_HUD,PERMIT_HUD, DNR_HUD) //SKYRAT EDIT ADDITION - PERMIT_HUD, DNR_HUD
```

File: code/modules/mob/living/living.dm [hunk 14]
```diff
@@ -24,6 +24,7 @@
 /mob/living/proc/prepare_data_huds()
 	med_hud_set_health()
 	med_hud_set_status()
+	update_dnr_hud() // SKYRAT EDIT ADDITION - DNR HUD
```
Comment (GoldenAlpharex) [inline code/modules/mob/living/living.dm]: Does it really need to be here, instead of having it modular? It's not like if extending this proc was really that hard to do modularly.

File: code/modules/mob/living/living_defines.dm [hunk 15]
```diff
@@ -1,6 +1,6 @@
 /mob/living
 	see_invisible = SEE_INVISIBLE_LIVING
-	hud_possible = list(HEALTH_HUD,STATUS_HUD,ANTAG_HUD)
+	hud_possible = list(HEALTH_HUD,STATUS_HUD,ANTAG_HUD,DNR_HUD) // SKYRAT EDIT ADDITION - DNR_HUD
```
Comment (GoldenAlpharex) [inline code/modules/mob/living/living_defines.dm]: Same thing about original
```suggestion
	hud_possible = list(HEALTH_HUD,STATUS_HUD,ANTAG_HUD, DNR_HUD) // SKYRAT EDIT ADDITION - DNR_HUD
```

File: modular_skyrat/master_files/code/datums/traits/neutral.dm [hunk 16]
```diff
@@ -1,5 +1,14 @@
...
 
+GLOBAL_VAR_INIT(DNR_trait_overlay, generate_DNR_trait_overlay())
+
+/proc/generate_DNR_trait_overlay()
+	RETURN_TYPE(/mutable_appearance)
+
+	var/mutable_appearance/DNR_trait_overlay = mutable_appearance('modular_skyrat/modules/indicators/icons/DNR_trait_overlay.dmi', "DNR", FLY_LAYER)
+	DNR_trait_overlay.appearance_flags = APPEARANCE_UI_IGNORE_ALPHA | KEEP_APART
+	return DNR_trait_overlay
+
```
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/datums/traits/neutral.dm:12]:
```suggestion


```

File: modular_skyrat/master_files/code/datums/traits/neutral.dm [hunk 17]
```diff
@@ -31,6 +40,41 @@
...
+	old_holder.update_dnr_hud()
+
+/mob/living/proc/update_dnr_hud()
+	var/image/dnr_holder = hud_list?[DNR_HUD]
+	if(isnull(dnr_holder))
+		return
+
+	var/icon/I = icon(icon, icon_state, dir)
+	dnr_holder.pixel_y = I.Height() - world.icon_size
```
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/datums/traits/neutral.dm]: Single letter variables bad, please fix that.

File: modular_skyrat/master_files/code/datums/traits/neutral.dm [hunk 18]
```diff
@@ -31,6 +40,41 @@
...
+	quirk_holder.update_dnr_hud()
+
+/datum/quirk/dnr/remove()
+	var/mob/living/old_holder = quirk_holder
+
+	. = ..()
+
+	old_holder.update_dnr_hud()
+
+/mob/living/proc/update_dnr_hud()
```
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/datums/traits/neutral.dm:63]: Document this please.

File: modular_skyrat/master_files/code/datums/traits/neutral.dm [hunk 19]
```diff
@@ -1,5 +1,14 @@
 #define TRAIT_HYDRA_HEADS "hydrahead" // We still dont have a centralised trait file
 
+GLOBAL_VAR_INIT(DNR_trait_overlay, generate_DNR_trait_overlay())
+
+/proc/generate_DNR_trait_overlay()
```
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/datums/traits/neutral.dm:6]: Document this please.

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
I'm not wholly convinced by the look of the icon, to be entirely honest, but I'm not entirely sure what the best plan of action would be for it either.

**Comment by nikothedude** [comment]
gahh i will fix this up later

File: code/modules/mob/living/living.dm (hunk 14 above)
Comment (nikothedude) [reply code/modules/mob/living/living.dm]: oh yeah it can be done modularly im just dumb

**Comment by nikothedude** [comment]
Also, I'd just like to say that I thought up a second, more valid rationale for this PR

Balance aside (whic hreally isnt that big a deal to begin with), seeing a DNR person dying is, in of itself, enough to intensify a situation and make things feel DANGEROUS. This is something that is sorely missed in our medical, and is missed by me in particular, a medical main.

Hell, this is something that is probably missed universally due to our medical. Whenever we see someone dying, it might get a bit more tense, but it never gets extremely scary unless they're about to blow up or something - and even then, if we have faith in medical, we can just shrug and say "oh well just SAD them". 
Ideally, DNR would change this. But nobody knows who is DNR and who isnt. The point of this PR, unbeknownst to me, was actually to just bring this intensity to light. When someone sees someone who dies forever slowly _dying_, it doesn't matter who they are, or what role they are, they will get a bit tenser as they realize something permanently _bad_ might happen.

**Review by GoldenAlpharex — APPROVED** [review]
